drivers/spi/spi_gd32.c

```diff
@@ -37,20 +41,74 @@ LOG_MODULE_REGISTER(spi_gd32);
 #else
 #error Unknown GD32 soc series
 #endif
+
+#ifdef CONFIG_SPI_GD32_DMA
+
+enum spi_gd32_dma_direction {
+	RX = 0,
+	TX,
+	NUM_OF_DIRECTION
+};
+
+struct spi_gd32_dma_config {
+	const struct device *dev;
+	uint32_t channel;
+	uint32_t config;
+	uint32_t slot;
+	uint32_t fifo_threshold;
+};
+
+struct spi_gd32_dma_data {
+	struct dma_config config;
+	struct dma_block_config block;
+	uint32_t count;
+};
+
+#endif
+
 struct spi_gd32_config {
 	uint32_t reg;
 	uint16_t clkid;
 	struct reset_dt_spec reset;
 	const struct pinctrl_dev_config *pcfg;
+#ifdef CONFIG_SPI_GD32_DMA
+	const struct spi_gd32_dma_config dma[NUM_OF_DIRECTION];
+#endif
 #ifdef CONFIG_SPI_GD32_INTERRUPT
 	void (*irq_configure)();
 #endif
 };
 
 struct spi_gd32_data {
 	struct spi_context ctx;
+#ifdef CONFIG_SPI_GD32_DMA
+	struct spi_gd32_dma_data dma[NUM_OF_DIRECTION];
+#endif
 };
 
+#ifdef CONFIG_SPI_GD32_DMA
+
+static uint32_t dummy_tx;
+static uint32_t dummy_rx;
+
+static bool spi_gd32_dma_enabled(const struct device *dev)
+{
+	const struct spi_gd32_config *cfg = dev->config;
+
+	if (cfg->dma[TX].dev && cfg->dma[RX].dev) {
+		return true;
+	}
+
+	return false;
+}
+
+static size_t spi_gd32_dma_enabled_num(const struct device *dev)
```

Comment:
Compared with `ARRAY_SIZE`, function call have an extra overhead.

Reply:
Determining from ARRAY_SIZE won't work if both exists DMA enabled SPI and disabled one.